File: alg8/alg8.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <algorithm>

using namespace std;

int n_vertex;
int m_edge;
int k_operation;

template<typename T>
ostream& operator<<(ostream& os, const vector<T>& v) {
    for(const T& t : v) {
        os << t;
    }

    return os << endl;
}

template<typename T>
ostream& operator<<(ostream& os, const vector<T*>& v) {
    for(const T *t : v) {
        os << *t;
    }

    return os << endl;
}

struct Vertex;
struct Command;
struct Component;

using Commandlist = vector<Command>;
using Vertexlist = vector<Vertex>;
using VertexPtrlist = vector<Vertex*>;
using Answerlist = vector<string>;

void setVertexPtrlistComponent(VertexPtrlist& lst, Component *cmp);

struct Component {
    VertexPtrlist vertex_ptrlist;

    Component(Vertex *v) : vertex_ptrlist{v} {}

    //TODO

    friend Component *merge(Component *cmp1, Component *cmp2) { 
        auto& lst1 = cmp1->vertex_ptrlist;
        auto& lst2 = cmp2->vertex_ptrlist;
        auto *cmp_dst = cmp1;
        auto *cmp_src = cmp2;

        if(lst1.size() > lst2.size()) {
            cmp_dst = cmp1;
            cmp_src = cmp2;
        } else {
            cmp_dst = cmp2;
            cmp_src = cmp1;
        }
        
        auto& lst_src = cmp_src->vertex_ptrlist;
        auto& lst_dst = cmp_dst->vertex_ptrlist;

        lst_dst.insert(lst_dst.end(), lst_src.begin(), lst_src.end());

        setVertexPtrlistComponent(lst_src, cmp_dst);

//        delete cmp_src;

        return cmp_dst;
    }
};

struct Vertex {
    int index {0};
    Component *component; 

    Vertex(int ii) : index{ii}, component{new Component{this}} {
//         cout << "vertex: " << component->vertex_ptrlist << endl;
    }

    friend ostream& operator<<(ostream& os, const Vertex& v) {
        os << v.index << " | " << v.component << endl;
        return os;
    }

    friend bool ask(const Vertex& u, const Vertex& v) {
        return u.component == v.component;
    }

    //TODO
    friend void addEdge(Vertex& u, Vertex& v) {
        Component *cmpU = u.component;
        Component *cmpV = v.component;

        try {
            cout << cmpU->vertex_ptrlist;
            cout << cmpV->vertex_ptrlist;

            Component *cmp = merge(cmpU, cmpV);

            cout << cmp->vertex_ptrlist;
            cout << "************************" << endl;
        } 
        catch(length_error) {
            cout << "Merge exception" << endl;
        }
    }
};

void setVertexPtrlistComponent(VertexPtrlist& lst, Component *cmp) {
    for(auto *pv : lst) {
        pv->component = cmp;
    }
}


struct Command {
    string name;
    int ui, vi; // индексы двух вершин, к которым относится команда

    friend ostream& operator<<(ostream& os, const Command& cmd) {
        os << cmd.name << " " << cmd.ui << " -> " << cmd.vi << endl;
        return os;
    }

    friend istream& operator>>(istream& is, Command& cmd) {
        return is >> cmd.name >> cmd.ui >> cmd.vi;
    }
};

Commandlist commandInit() {
    string filename = "data.txt";
    ifstream fin{filename};

    fin >> n_vertex >> m_edge >> k_operation;
    cout << n_vertex << " " << m_edge << " " << k_operation << endl;

    for(int i = 0; i < m_edge; i++) {
        int ui, vi;

        fin >> ui >> vi;
    }

    Commandlist commands;

    for(int i = 0; i < k_operation; i++) {
        Command cmd;
        fin >> cmd;

        commands.push_back(cmd);
    }

    reverse(commands.begin(), commands.end());

    return commands;
}

Vertexlist vertexInit() {
    Vertexlist vertexlist;
    vertexlist.emplace_back(0);

        cout << "vertexInit0: " << vertexlist[0].component->vertex_ptrlist;
    for(int vi = 1; vi <= n_vertex; vi++) {
        vertexlist.emplace_back(vi);
    }
    
    for(auto& v : vertexlist) {
        cout << "vertexInit: " << v.component->vertex_ptrlist;
    }

    return vertexlist;
}

Answerlist execute(Commandlist& reversed_commandlist, Vertexlist& vertexlist) {
    Answerlist anslst;

    for(auto cmd : reversed_commandlist) {
        Vertex& u = vertexlist[cmd.ui];
        Vertex& v = vertexlist[cmd.vi];

        if(cmd.name == "cut") {
            addEdge(u, v);
        } else {
            anslst.push_back(ask(u, v) ? "YES\n" : "NO\n");
        }
    }

    return anslst;
}

int main() {
    Commandlist reversed_commandlist = commandInit();
    cout << reversed_commandlist;

    Vertexlist vertexlist = vertexInit();
    Answerlist answerlist = execute(reversed_commandlist, vertexlist);

    cout << answerlist;

} 

File: alg8/alg8.py
class Component:
    def __init__(self, vertex):
        self.vertex_list = [vertex]

def merge(cmp1, cmp2):
    if cmp1 == cmp2:
        return cmp1
    
    cmp_dst, cmp_src = (cmp1, cmp2) if len(cmp1.vertex_list) > len(cmp2.vertex_list) else (cmp2, cmp1)
    
    lst_src = cmp_src.vertex_list
    lst_dst = cmp_dst.vertex_list
    lst_dst += lst_src

    set_vertex_list_component(lst_src, cmp_dst)

    return cmp_dst

class Vertex:
    def __init__(self, ii):
        self.index = ii
        self.component = Component(self)

    def __repr__(self):
        return str(self.index) + " | " + str(self.component)
    
def ask(u, v):
    return u.component == v.component

def add_edge(u, v):
    cmpU, cmpV = u.component, v.component
    print(cmpU.vertex_list)
    print(cmpV.vertex_list)
    cmp = merge(cmpU, cmpV)
    print(cmp.vertex_list)
    print("-------------------")

class Command:
    def __init__(self, name, ui, vi):
        self.name = name
        self.ui = int(ui)
        self.vi = int(vi)

    def __repr__(self):
        return self.name + " " + str(self.ui) + " -> " + str(self.vi)

def print_list(lst):
    print(*lst, sep = "\n")

def set_vertex_list_component(lst, cmp):
    for v in lst:
        v.component = cmp

def command_init():
    filename = "data.txt"
    fin = open(filename, 'r')
    n_vertex, m_edge, k_operation = map(int, fin.readline().split())
    print(n_vertex, m_edge, k_operation)
    
    for _ in range(m_edge):
        ui, vi = map(int, fin.readline().split())
    
    commands = []
    for _ in range(k_operation):
        cmd = Command(*fin.readline().split())
        commands.append(cmd)

    fin.close()
    commands.reverse()
    return commands

def vertex_init(n_vertex):
    return [Vertex(vi) for vi in range(n_vertex + 1)]

def execute(reversed_commandlist, vertex_list):
    anslst = []
    for cmd in reversed_commandlist:
        u, v = vertex_list[cmd.ui], vertex_list[cmd.vi]
        if cmd.name == "cut":
            add_edge(u, v)
        else:
            anslst.append("YES\n" if ask(u, v) else "NO\n")
    
    return reversed(anslst)

if __name__ == "__main__":
    reversed_commandlist = command_init()
    print_list(reversed_commandlist)

    vertex_list = vertex_init(reversed_commandlist[0].ui)
    answerlist = execute(reversed_commandlist, vertex_list)
    print_list(answerlist)

